server/backup_runner.go

```diff
@@ -70,6 +73,90 @@ func (r *BackupRunner) Run(ctx context.Context, wg *sync.WaitGroup) {
 	}
 }
 
+// TODO(dragonly): Make best effort to assure that users could recover to at least RetentionPeriodTs ago.
+// This may require pending deleting expired backup files and binlog files.
+func (r *BackupRunner) purgeExpiredBackupData(ctx context.Context) {
+	backupSettingList, err := r.server.store.FindBackupSetting(ctx, api.BackupSettingFind{})
+	if err != nil {
+		log.Error("Failed to find all the backup settings", zap.Error(err))
+		return
+	}
+
+	log.Debug("Deleting expired snapshot backup files")
+	for _, bs := range backupSettingList {
+		if bs.RetentionPeriodTs == api.BackupRetentionPeriodUnset {
+			continue
+		}
+		backupDir := path.Join(r.server.profile.DataDir, getBackupRelativeDir(bs.DatabaseID))
+		backupFileInfoList, err := ioutil.ReadDir(backupDir)
+		if err != nil {
+			log.Error("Failed to read backup directory", zap.String("path", backupDir), zap.Error(err))
+			continue
+		}
+		for _, backupFileInfo := range backupFileInfoList {
+			expireTime := backupFileInfo.ModTime().Add(time.Duration(bs.RetentionPeriodTs))
+			if time.Now().After(expireTime) {
+				backupFilePath := path.Join(backupDir, backupFileInfo.Name())
+				if err := os.Remove(backupFilePath); err != nil {
+					log.Error("Failed to remove an expired backup file", zap.String("path", backupFilePath), zap.Error(err))
+				}
+				log.Info("Deleted expired backup file", zap.String("path", backupFilePath))
+			}
+		}
+	}
+
+	log.Debug("Deleting expired MySQL binlog files")
+	normal := api.Normal
+	instanceList, err := r.server.store.FindInstance(ctx, &api.InstanceFind{RowStatus: &normal})
+	if err != nil {
+		log.Error("Failed to find non-archived instances", zap.Error(err))
+		return
+	}
+
+	for _, instance := range instanceList {
+		if instance.Engine != db.MySQL {
+			continue
+		}
+		backupSettingList, err := r.server.store.FindBackupSetting(ctx, api.BackupSettingFind{InstanceID: &instance.ID})
+		if err != nil {
+			log.Error("Failed to find backup settings for instance", zap.String("instance", instance.Name), zap.Error(err))
+			continue // next instance
+		}
+		maxRetentionPeriodForInstance := api.BackupRetentionPeriodUnset
```

Comment:
If users don't set retention period in the backup settings, do they mean never to expire any backups?

Reply:
The logic looks strange.

maxRetentionPeriodForInstance := 0 // This looks strange, why zero.

I would consider to do the following.

```
const RetentionPeriodUnset = 0

maxRetentionPeriodForInstance := maxInt
for ... {
  if period != RetentionPeriodUnset && period < maxRetentionPeriodForInstance {
    maxRetentionPeriodForInstance = period
  }
}
```

Reply:
Ah, I see... It's not clever to set an initial max value to 0 😂